MapleServer2/Managers/Actors/Npc.cs

```diff
@@ -1,30 +1,415 @@
     public MobMovement Movement { get; set; }
+    public IFieldActor<Player> Target;
+
+    private CoordS NextMovementTarget;
+    private float Distance;
+
+    private int LastMovementTime;
+    private float BaseVelocity => Movement is MobMovement.Follow or MobMovement.Run ? Value.NpcMetadataSpeed.RunSpeed : Value.NpcMetadataSpeed.WalkSpeed;
 
-    public Npc(int objectId, int npcId, FieldManager fieldManager) : this(objectId, NpcMetadataStorage.GetNpcMetadata(npcId), fieldManager) { }
+    private PatrolData PatrolData;
+    public Queue<WayPoint> WayPointQueue = new();
+    private WayPoint CurrentWayPoint;
+    private bool WaitForAnimation;
+
+    public Npc(int objectId, int mobId, FieldManager fieldManager) : this(objectId, NpcMetadataStorage.GetNpcMetadata(mobId), fieldManager) { }
 
     public Npc(int objectId, NpcMetadata metadata, FieldManager fieldManager) : base(objectId, metadata, fieldManager)
     {
-        Animation = 255;
+        Animation = AnimationStorage.GetSequenceIdBySequenceName(metadata.Model, "Idle_A");
+        AI = MobAIManager.GetAI(metadata.AiInfo);
         Stats = new(metadata);
         State = NpcState.Normal;
         Action = NpcAction.Idle;
     }
 
+    public void Attack()
+    {
+        int roll = Rand.Next(100);
+        for (int i = 0; i < Value.NpcMetadataSkill.SkillIds.Length; i++)
+        {
+            if (roll < Value.NpcMetadataSkill.SkillProbs[i])
+            {
+                // Rolled this skill.
+                Cast(new(Value.NpcMetadataSkill.SkillIds[i], Value.NpcMetadataSkill.SkillLevels[i]));
+                StartSkillTimer((Value.NpcMetadataSkill.SkillCooldown > 0) ? Value.NpcMetadataSkill.SkillCooldown : 1000);
+            }
+
+            roll -= Value.NpcMetadataSkill.SkillProbs[i];
+        }
+    }
+
     public override void Animate(string sequenceName)
     {
         Animation = AnimationStorage.GetSequenceIdBySequenceName(Value.Model, sequenceName);
         // TODO implement stopping animation
     }
+
+    public override void Damage(DamageHandler damage, GameSession session)
+    {
+        if (Value.Friendly == 2) // Attacking friendly NPC
+        {
+            return;
+        }
+
+        base.Damage(damage, session);
+
+        session.FieldManager.BroadcastPacket(StatPacket.UpdateStats(this, StatAttribute.Hp));
+        if (IsDead)
+        {
+            HandleMobKill(session, this);
+        }
+    }
+
+    public override void Perish()
+    {
+        IsDead = true;
+        State = NpcState.Dead;
+        Animate(Value.NpcMetadataDead.Actions.ElementAtOrDefault(0));
+        Velocity = default;
+    }
+
+    /// <summary>
+    /// Update the NPC's velocity and LookDirection.
+    /// </summary>
+    /// <returns>True if the NPC is moving.</returns>
+    public bool UpdateVelocity()
+    {
+        if (WayPointQueue.Count > 0 && Distance <= 0)
+        {
+            return GetNextPath();
+        }
+
+        if (Distance <= 0)
```

Comment:
Rather than check `Distance <= 0` twice, try checking it once, then immediately check the `Count`, if it's greater than `0`, return that, otherwise return false.